docs/lang/articles/visualization/ggui.md

```diff
@@ -108,6 +109,212 @@ If a mesh has `num` triangles, the `indices` should be a 1D scalar field with a
+points_pos = ti.Vector.field(3, dtype=ti.f32, shape = N)
+points_indices = ti.Vector.field(1, dtype=ti.i32, shape = N)
+
+@ti.kernel
+def init_points_pos(points : ti.template()):
+    for i in range(points.shape[0]):
+        points[i] = [i for j in range(3)]
+        # points[i] = [ti.sin(i * 1.0), i * 0.2, ti.cos(i * 1.0)]
+
+@ti.kernel
+def init_points_indices(points_indices : ti.template()):
+    for i in range(N):
+        points_indices[i][0] = i // 2 + i % 2
+
+init_points_pos(particles_pos)
+init_points_pos(points_pos)
+init_points_indices(points_indices)
+
+window = ti.ui.Window("Test for Drawing 3d-lines", (768, 768))
+canvas = window.get_canvas()
+scene = ti.ui.Scene()
+camera = ti.ui.make_camera()
+camera.position(5, 2, 2)
+
+while window.running:
+    camera.track_user_inputs(window, movement_speed=0.03, hold_key=ti.ui.RMB)
+    scene.set_camera(camera)
+    scene.ambient_light((0.8, 0.8, 0.8))
+    scene.point_light(pos=(0.5, 1.5, 1.5), color=(1, 1, 1))
+
+    scene.particles(particles_pos, color = (0.68, 0.26, 0.19), radius = 0.1)
+    # Here you will get visible part from the 3rd point with (N - 4) points.
+    scene.lines(points_pos, color = (0.28, 0.68, 0.99), width = 5.0, vertex_count = N - 4, vertex_offset = 2)
+    # Using indices to indicate which vertex to use
+    # scene.lines(points_pos, color = (0.28, 0.68, 0.99), width = 5.0, indices = points_indices)
+    # Case 1, vertex_count will be changed to N - 2 when drawing.
+    # scene.lines(points_pos, color = (0.28, 0.68, 0.99), width = 5.0, vertex_count = N - 1, vertex_offset = 0)
+    # Case 2, vertex_count will be changed to N - 2 when drawing.
+    # scene.lines(points_pos, color = (0.28, 0.68, 0.99), width = 5.0, vertex_count = N, vertex_offset = 2)
+    canvas.scene(scene)
+    window.show()
+```
+
+3. Details of mesh instancing
+```python
+num_instance  = 100
+m_transforms = ti.Matrix.field(4, 4, dtype = ti.f32, shape = num_instance)
+
+
+# For example: An object is scaled by 2, rotated by rotMat, and translated by t = [1, 2, 3], then
+#
+# The ScaleMatrix is:
+# 2, 0, 0, 0
+# 0, 2, 0, 0
+# 0, 0, 2, 0
+# 0, 0, 0, 1
+#
+# The RotationMatrix is:
+# https://en.wikipedia.org/wiki/Rotation_matrix#General_rotations
+#
+# The TranslationMatrix is:
+# 1, 0, 0, 1
+# 0, 1, 0, 2
+# 0, 0, 1, 3
+# 0, 0, 0, 1
+#
+# Let TransformMatrix = TranslationMatrix @ RotationMatrix @ ScaleMatrix, then the final TransformMatrix is:
+#   2 * rotMat00,     rotMat01,       rotMat02, 1
+#       rotMat10, 2 * rotMat11,       rotMat12, 2
+#       rotMat20,     rotMat21,   2 * rotMat22, 3
+#              0,            0,              0, 1
+...
+
+# Here draw mesh instances (will draw from the 1st instance)
+scene.mesh_instance(vertices, indices, transforms = m_transforms, instance_offset = 1)
+```
+4. Example of setting wareframe mode
+```python
+
+window = ti.ui.Window("Display Mesh", (1024, 1024), vsync=True)
+canvas = window.get_canvas()
+scene = ti.ui.Scene()
+camera = ti.ui.make_camera()
```

Comment:
i thought `ti.ui.Camera` can be directly constructed. it should be on the wish list I wrote before #5426 .

Reply:
well, fine, i thought you took over this. lets fix this a bit later